pysmt/rewritings.py

```diff
@@ -0,0 +1,353 @@
+            return mgr.And(memo[x] for x in formula.args())
+
+        elif formula.is_or():
+            return mgr.Or(memo[x] for x in formula.args())
+
+        elif formula.is_ite():
+            # This must be a boolean ITE as we do not recur within
+            # theory atoms
+            assert self.env.stc.get_type(formula).is_bool_type()
+            i, t, e = formula.args()
+            ni = mgr.Not(i)
+            return mgr.And(mgr.Or(memo[ni], memo[t]),
+                           mgr.Or(memo[i], memo[e]))
+
+        elif formula.is_forall():
+            return mgr.ForAll(formula.quantifier_vars(), memo[formula.arg(0)])
+
+        elif formula.is_exists():
+            return mgr.Exists(formula.quantifier_vars(), memo[formula.arg(0)])
+
+        else:
+            # This is a theory atom
+            return formula
+
+
+class PrenexNormalizer(DagWalker):
+    """
+    This class traverses a formula and rebuilds it in prenex normal form.
+    """
+
+    def __init__(self, env=None, invalidate_memoization=None):
+        DagWalker.__init__(self,
+                           env=env,
+                           invalidate_memoization=invalidate_memoization)
+        self.mgr = self.env.formula_manager
+        self.check_symbol = self.mgr.FreshSymbol(types.BOOL)
+
+        # The walker returns a pair (L, m) where m is a
+        # quantifier-free formula (the matrix) and L is a list of
+        # pairs (Q, vars) where Q is either mgr.Exists or mgr.ForAll
+        # and vars is a frozenset of variables.  The semantics is that
+        # the input formula is equivalent to res computed as follows:
+        # res = m
+        # for Q, vars in L:
+        #    res = Q(vars, res)
+
+        for elem in op.ALL_TYPES:
+            if elem in op.QUANTIFIERS:
+                self.functions[elem] = self.walk_quantifier
+            elif elem in op.CONSTANTS:
+                self.functions[elem] = self.walk_constant
+            elif elem in op.RELATIONS:
+                self.functions[elem] = self.walk_theory_relation
+            elif elem in op.BV_OPERATORS or elem in op.LIRA_OPERATORS:
+                self.functions[elem] = self.walk_theory_op
+            else:
+                self.functions[elem] = self.walk_error
+        self.functions[op.SYMBOL] = self.walk_symbol
+        self.functions[op.FUNCTION] = self.walk_function
+        self.functions[op.ITE] = self.walk_ite
+        self.functions[op.AND] = self.walk_conj_disj
+        self.functions[op.OR] = self.walk_conj_disj
+        self.functions[op.NOT] = self.walk_not
+        self.functions[op.IFF] = self.walk_iff
+        self.functions[op.IMPLIES] = self.walk_implies
+
+
+    def normalize(self, formula):
+        quantifiers, matrix = self.walk(formula)
+        res = matrix
+        for Q, qvars in quantifiers:
+            res = Q(qvars, res)
+        return res
+
+    def _invert_quantifier(self, Q):
+        if Q([self.check_symbol], self.check_symbol).is_exists():
```

Comment:
This seems a bit baroque... would it suffice to compare with the constructor Exists() ?